node/src/Pos/Block/Logic/Header.hs

```diff
@@ -328,35 +327,89 @@ getHeadersOlderExp upto = do
                 | otherwise = selGo es ii $ succ skipped
         in selGo elems ixs 0
 
--- CSL-396 don't load all the blocks into memory at once
--- | Given @from@ and @to@ headers where @from@ is older (not strict)
--- than @to@, and valid chain in between can be found, headers in
--- range @[from..to]@ will be found.
-getHeadersFromToIncl
-    :: forall ssc m .
-       (HasConfiguration, MonadDBRead m, SscHelpersClass ssc)
-    => HeaderHash -> HeaderHash -> m (Maybe (OldestFirst NE HeaderHash))
-getHeadersFromToIncl older newer = runMaybeT . fmap OldestFirst $ do
+-- | Given optional @depthLimit@, @from@ and @to@ headers where @from@
+-- is older (not strict) than @to@, and valid chain in between can be
+-- found, headers in range @[from..to]@ will be found. If the number
+-- of headers in the chain (which should be returned) is more than
+-- @depthLimit@, error will be thrown.
+getHeadersRange ::
+       forall ssc m. (HasConfiguration, MonadDBRead m, SscHelpersClass ssc)
+    => Maybe Word
+    -> HeaderHash
+    -> HeaderHash
+    -> m (Either Text (OldestFirst NE HeaderHash))
+getHeadersRange _ older newer | older == newer = runExceptT $ do
+    unlessM (isJust <$> (DB.blkGetHeader @ssc newer)) $
+        throwError "getHeadersRange: can't find newer-older header"
+    pure $ OldestFirst $ one newer
+getHeadersRange depthLimitM older newer = runExceptT $ do
     -- oldest and newest blocks do exist
-    start <- MaybeT $ DB.blkGetHeader @ssc newer
-    end   <- MaybeT $ DB.blkGetHeader @ssc older
-    guard $ getEpochOrSlot start >= getEpochOrSlot end
-    let lowerBound = getEpochOrSlot end
-    if newer == older
-    then pure $ one newer
-    else loadHeadersDo lowerBound (one newer) $ start ^. prevBlockL
+    newerHd <- fromMaybeM "can't retrieve newer header" $ DB.blkGetHeader @ssc newer
+    olderHd <- fromMaybeM "can't retrieve older header" $ DB.blkGetHeader @ssc older
+    let olderD = olderHd ^. difficultyL
+    let newerD = newerHd ^. difficultyL
```

Comment:
What's the significance of using difficulty rather than `getEpochOrSlot` here?

Reply:
The particular function is operating in terms of difficulty. We can't give more than `depthLimit` blocks at once, so even if I used `EpochOrSlot` (which i actually did in the beginning), I would need to convert it to number of blocks somehow (for example, counting difficulty in traversal).